File: android/src/main/java/co/nvapps/googleump/UmpModule.kt
package co.nvapps.googleump

import android.content.Context
import com.google.android.ump.ConsentDebugSettings
import com.google.android.ump.ConsentInformation
import com.google.android.ump.ConsentRequestParameters
import com.google.android.ump.UserMessagingPlatform
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition


class UmpModule : Module() {

  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  private val currentActivity
    get() = appContext.activityProvider?.currentActivity ?: throw Exceptions.MissingActivity()

  private lateinit var consentInformation: ConsentInformation

  override fun definition() = ModuleDefinition {
    Name("Ump")

    OnCreate {
      consentInformation = UserMessagingPlatform.getConsentInformation(context)
    }

    Function("getConsentInformation") {
      return@Function UmpConsentInformation.fromConsentInformation(consentInformation)
    }

    AsyncFunction("requestInfoUpdate") { options: UmpConsentRequestParameters, promise: Promise ->
      try {
        val consentRequestParameters = ConsentRequestParameters.Builder().apply {
          setTagForUnderAgeOfConsent(options.tagForUnderAgeOfConsent)
          options.debugSettings?.let {
            val debugSettingsBuilder = ConsentDebugSettings.Builder(context).apply {
              it.testDeviceIdentifiers?.forEach { id ->
                addTestDeviceHashedId(id)
              }
              it.debugGeography?.let {
                setDebugGeography(it.toDebugGeography())
              }
            }
            setConsentDebugSettings(debugSettingsBuilder.build())
          }
        }.build()

        consentInformation.requestConsentInfoUpdate(
          currentActivity,
          consentRequestParameters,
          { promise.resolve(UmpConsentInformation.fromConsentInformation(consentInformation)) },
          { formError ->
            promise.reject("E_FORM_ERROR", formError.message, null)
          })
      } catch (e: Exception) {
        promise.reject("E_UNKNOWN_ERROR", e.message, e)
      }
    }

    AsyncFunction("loadAndShowConsentFormIfRequired") { promise: Promise ->
      try {
        currentActivity.runOnUiThread {
          UserMessagingPlatform.loadAndShowConsentFormIfRequired(currentActivity) { loadAndShowError ->
            if (loadAndShowError != null) {
              promise.reject(
                "E_LOAD_AND_SHOW_ERROR",
                "Error Code: ${loadAndShowError.errorCode}, Message: ${loadAndShowError.message}",
                null
              )
            } else {
              promise.resolve(UmpConsentInformation.fromConsentInformation(consentInformation))
            }
          }
        }
      } catch (e: Exception) {
        promise.reject("E_UNKNOWN_ERROR", e.message ?: "Unknown Error", e)
      }
    }

    AsyncFunction("showPrivacyOptionsForm") { promise: Promise ->
      try {
        currentActivity.runOnUiThread {
          UserMessagingPlatform.showPrivacyOptionsForm(currentActivity) { formError ->
            if (formError != null) {
              promise.reject(
                "E_FORM_ERROR",
                "Error Code: ${formError.errorCode}, Message: ${formError.message}",
                null
              )
            } else {
              promise.resolve(UmpConsentInformation.fromConsentInformation(consentInformation))
            }
          }
        }
      } catch (e: Exception) {
        promise.reject("E_UNKNOWN_ERROR", e.message ?: "Unknown Error", e)
      }
    }

    Function("reset") {
      consentInformation.reset()
    }
  }
}

File: ios/UmpModule.swift
import ExpoModulesCore
import UserMessagingPlatform

final class MissingViewControllerException: Exception {
    override var reason: String {
        "Unable to find the current view controller to present the UMP form"
    }
}

public class UmpModule: Module {
    public func definition() -> ModuleDefinition {
        Name("Ump")

        Function("getConsentInformation") {
            return getConsentInformation()
        }

        AsyncFunction("requestInfoUpdate") { (requestParameters: RequestParameters, promise: Promise) in
            let umpRequestParameters = UMPRequestParameters()

            umpRequestParameters.tagForUnderAgeOfConsent = requestParameters.tagForUnderAgeOfConsent

            if let debugSettings = requestParameters.debugSettings {
                let umpDebugSettings = UMPDebugSettings()

                if debugSettings.testDeviceIdentifiers != nil {
                    umpDebugSettings.testDeviceIdentifiers = debugSettings.testDeviceIdentifiers
                }

                if let debugGeography = debugSettings.debugGeography {
                    umpDebugSettings.geography = debugGeography.toUMPDebugGeography()
                }

                umpRequestParameters.debugSettings = umpDebugSettings
            }

            UMPConsentInformation.sharedInstance.requestConsentInfoUpdate(
                with: umpRequestParameters,
                completionHandler: { error in self.handleResult(error: error, promise: promise) })
        }

        AsyncFunction("loadAndShowConsentFormIfRequired", { (promise: Promise) in
            guard let currentVc = appContext?.utilities?.currentViewController() else {
                throw MissingViewControllerException()
            }

            UMPConsentForm.loadAndPresentIfRequired(from: currentVc) { error in
                self.handleResult(error: error, promise: promise)
            }
        }).runOnQueue(DispatchQueue.main)

        AsyncFunction("showPrivacyOptionsForm", { (promise: Promise) in
            guard let currentVc = appContext?.utilities?.currentViewController() else {
                throw MissingViewControllerException()
            }

            UMPConsentForm.presentPrivacyOptionsForm(from: currentVc) { error in
                self.handleResult(error: error, promise: promise)
            }
        }).runOnQueue(DispatchQueue.main)

        Function("reset") {
            UMPConsentInformation.sharedInstance.reset()
        }
    }

    private func getConsentInformation() -> ConsentInformation {
        return ConsentInformation.fromUMPConsentInformation(consentInformation: UMPConsentInformation.sharedInstance)
    }

    private func handleResult(error: Error?, promise: Promise) {
        if let error = error {
            promise.reject("E_FORM_ERROR", error.localizedDescription)
        } else {
            promise.resolve(getConsentInformation())
        }
    }
}
